plugins/ccgx/fu-ccgx-dmc-firmware.c

```diff
@@ -0,0 +1,332 @@
+	g_autofree guint8 *fwct_buff = NULL;
+	g_autofree guint8 *custom_meta_buff = NULL;
+
+	/* read fwct info */
+	if (!fu_memcpy_safe ((guint8 *) &self->fwct_info, sizeof(self->fwct_info), 0x0, /* dst */
+			     fw_buff, fw_buff_size, 0, sizeof(self->fwct_info), error)) /* src */
+		return FALSE;
+	/* check for 'F' 'W' 'C' 'T' in signature */
+	if (self->fwct_info.signature != DMC_FWCT_SIGN) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid dmc signature, expected 0x%04X got 0x%04X",
+			     (guint32)DMC_FWCT_SIGN, (guint32)self->fwct_info.signature);
+		return FALSE;
+	}
+	/* check fwct size */
+	if (self->fwct_info.size > DMC_FWCT_MAX_SIZE || self->fwct_info.size == 0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "invalid dmc fwct size, expected equal to or less than %u, got %u",
+			     (guint32)DMC_FWCT_MAX_SIZE,  (guint32)self->fwct_info.size);
+		return FALSE;
+	}
+	/* alloc fwct buffer */
+	fwct_buff = g_malloc0 (self->fwct_info.size);
+	/* read fwct data */
+	if (!fu_memcpy_safe ((guint8 *)fwct_buff, self->fwct_info.size, 0x0, /* dst */
+		     fw_buff, fw_buff_size, 0, self->fwct_info.size, error)) /* src */
+		return FALSE;
+	/* create fwct binary */
+	self->fwct_record = g_bytes_new (fwct_buff, self->fwct_info.size);
+	/* create custom meta binary */
+	if (!fu_common_read_uint16_safe (fw_buff, fw_buff_size, self->fwct_info.size, &custom_meta_size,
+			    G_LITTLE_ENDIAN,error))
+		return FALSE;
+	if (custom_meta_size > 0) {
+		/* alloc custom meta buffer */
+		custom_meta_buff = g_malloc0 (custom_meta_size);
+		/* read custom meta data */
+		if (!fu_memcpy_safe ((guint8 *)custom_meta_buff, custom_meta_size, 0x0, /* dst */
+			     fw_buff, fw_buff_size, self->fwct_info.size + 2, custom_meta_size, error)) /* src */
+			return FALSE;
+		self->custom_meta_record = g_bytes_new (custom_meta_buff, custom_meta_size);
+	}
+
+	/* set row data start offset */
+	self->row_data_offset_start = self->fwct_info.size + DMC_CUSTOM_META_LENGTH_FIELD_SIZE + custom_meta_size;
+	self->fw_data_size = fw_buff_size - self->row_data_offset_start;
+	/* parse image */
+	if (!fu_ccgx_dmc_firmware_parse_image (firmware, fw_buff, fw_buff_size, error))
+		return FALSE;
+
+	/* add something, although we'll use the records for the update */
+	fu_firmware_image_set_addr (img, 0);
+	fu_firmware_add_image (firmware, img);
+	return TRUE;
+}
+
+static void
+fu_ccgx_dmc_firmware_init (FuCcgxDmcFirmware *self)
+{
+	self->image_records = g_ptr_array_new_with_free_func ((GFreeFunc) fu_ccgx_dmc_firmware_image_record_free);
+}
+
+static void
+fu_ccgx_dmc_firmware_finalize (GObject *object)
+{
+	FuCcgxDmcFirmware *self = FU_CCGX_DMC_FIRMWARE (object);
+	if (self->fwct_record)
+		g_bytes_unref (self->fwct_record);
+	if (self->custom_meta_record)
+		g_bytes_unref (self->custom_meta_record);
+	if (self->image_records)
+		g_ptr_array_unref (self->image_records);
```

Comment:
please use ` != NULL` when comparing pointers

Reply:
Fixed on today commit (ccgx: fix code alignment, modify code for review)